src/puya/compile.py

```diff
@@ -124,66 +153,110 @@ def get_mypy_options() -> mypy.options.Options:
+        except KeyError:
+            raise CodeError(f"invalid reference: {ref}") from None
+        if kind == "logic_sig" and isinstance(comp_ref, _CompiledLogicSig):
+            return assemble_program(context, comp_ref.program.teal, template_constants).bytecode
+        elif kind in ("approval", "clear_state") and isinstance(comp_ref, _CompiledContract):
+            program = comp_ref.approval_program if kind == "approval" else comp_ref.clear_program
+            return assemble_program(context, program.teal, template_constants).bytecode
         else:
-            for artifact_ir in module_ir:
-                out_dir = determine_out_dir(src.path.parent, context.options)
-                name = artifact_ir.metadata.name
-                artifact_ir_base_path = out_dir / name
-                artifact_ir = optimize_and_destructure_ir(
-                    context, artifact_ir, artifact_ir_base_path
-                )
+            raise InternalError(f"invalid kind: {kind}, {type(comp_ref)}")
 
-                result.setdefault(src, []).append(artifact_ir)
-    return result
+    all_artifact_irs = {
+        ir.metadata.ref: Artifact(
+            path=Path(ir.source_location.file),
+            ir=ir,
+        )
+        for module_ir in module_irs.values()
+        for ir in module_ir
+    }
 
+    optimize_context = attrs_extend(
+        IROptimizeContext,
+        context,
+        get_program_bytecode=get_program_bytecode,
+        state_totals={
+            artifact.ir.metadata.ref: artifact.ir.metadata.state_totals
+            for artifact in all_artifact_irs.values()
+            if isinstance(artifact.ir, ContractIR)
+        },
+    )
 
-def module_irs_to_teal(
-    context: CompileContext, module_irs: dict[ParseSource, list[ModuleArtifact]]
-) -> dict[ParseSource, list[CompilationArtifact]]:
-    result = dict[ParseSource, list[CompilationArtifact]]()
-    # used to check for conflicts that would occur on output
-    artifacts_by_output_base = dict[Path, ModuleArtifact]()
-    for src, artifact_irs in module_irs.items():
-        for artifact_ir in artifact_irs:
-            out_dir = determine_out_dir(src.path.parent, context.options)
-            name = artifact_ir.metadata.name
-            artifact_ir_base_path = out_dir / name
-            if existing := artifacts_by_output_base.get(artifact_ir_base_path):
-                logger.error(
-                    f"Duplicate contract name {name}", location=artifact_ir.source_location
+    for artifact in ArtifactCompilationSorter.sort(
+        [ref for refs in artifact_refs.values() for ref in refs], all_artifact_irs
+    ):
+        out_dir = determine_out_dir(artifact.path.parent, context.options)
+        name = artifact.ir.metadata.name
+        artifact_ir_base_path = out_dir / name
+
+        artifact_ir = optimize_and_destructure_ir(
+            optimize_context, artifact.ir, artifact_ir_base_path
+        )
+        # IR validation may reveal further errors
```

Comment:
could you expand this comment? and also mention why it's an exit rather than continue to next (possibly after inserting a dummy value for the result of the current one)?